sdk/nodejs/appservice/zMixins.ts

```diff
@@ -501,3 +501,264 @@ export function getResourceGroupNameAndLocation(
+        throw new pulumi.RunError("Cannot provide both [callback] and [callbackFactory]");
+    }
+
+    if (!args.callback && !args.callbackFactory) {
+        throw new Error("One of [callback] or [callbackFactory] must be provided.");
+    }
+
+    const func = args.callback
+        ? redirectConsoleOutput(args.callback)
+        : () => redirectConsoleOutput(args.callbackFactory!());
+
+    return pulumi.output(pulumi.runtime.serializeFunction(func, { isFactoryFunction: !!args.callbackFactory }));
+}
+
+export interface AzureFunctionArgs {
+    name: string;
+    bindings: BindingDefinition[];
+    func: pulumi.runtime.SerializedFunction;
+    appSettings?: { [key: string]: string };
+}
+
+// TODO: Should we inherit from pulumi.ComponentResource? If so, we should probably set parent to the Function App, but
+// how do we do so if Functions are created before the app (see the sample).
+export abstract class AzureFunction {
+    public readonly definition: pulumi.Output<AzureFunctionArgs>;
+
+    constructor(definition: pulumi.Output<AzureFunctionArgs>) {
+
+        this.definition = definition;
+    }
+}
+
+export interface AzureFunctionOutputBinding {
+    readonly definition: pulumi.Output<BindingDefinition>;
+    readonly appSettings: pulumi.Output<{ [key: string]: string }>;
+}
+
+function produceDeploymentPackage(args: MultiFunctionAppArgs): pulumi.Output<pulumi.asset.AssetMap> {
+    
+    if (args.functions === undefined) {
+        throw new Error("Can not serialize undefined Azure Functions");
+    }
+
+    return pulumi.all(args.functions.map(f => f.definition)).apply(async functions => {
+
+        const map: pulumi.asset.AssetMap = {};
+        map["host.json"] = new pulumi.asset.StringAsset(JSON.stringify({
+            version: "2.0",
+            tracing: { consoleLevel: "verbose" },
+            extensionBundle: {
+                id: "Microsoft.Azure.Functions.ExtensionBundle",
+                version: "[1.*, 2.0.0)"
+            },
+            ...args.hostSettings,
+        }));
+
+        const pathSet = await pulumi.runtime.computeCodePaths(args.codePathOptions);
+        for (const [path, value] of pathSet.entries()) {
+            map[path] = value;
+        }
+
+        for (const func of functions) {
+            map[`${func.name}/function.json`] = new pulumi.asset.StringAsset(JSON.stringify({
+                disabled: false,
+                bindings: func.bindings,
+            }));
+
+            map[`${func.name}/index.js`] = new pulumi.asset.StringAsset(`module.exports = require("./handler").handler`),
+            map[`${func.name}/handler.js`] = new pulumi.asset.StringAsset(func.func.text);
+        }
+
+        return map;
+    });
+}
+
+function combineAppSettings(args: MultiFunctionAppArgs): pulumi.Output<{[key: string]: string}> {
+    
+    const perFunctionSettings = args.functions !== undefined ? args.functions.map(c => c.definition.apply(b => b.appSettings)) : [];

```

Comment:
would it ever make sense to have a MultiFunctionApp without 'args.functions'?  if not, we should just make it non-optional.

Reply:
An alternative is to provide an Archive of a folder on the disk (e.g. for .NET/Java functions), and yet avoid setting Plan/Storage/Blob/settings manually.